core/trino-main/src/main/java/io/trino/operator/exchange/PartitioningExchanger.java

```diff
@@ -75,15 +75,19 @@ private synchronized void partitionPage(Page page, Page partitionPage)
             if (partitionSize == 0) {
                 continue;
             }
+            // reset the assigned positions list size for the next iteration to start empty
+            positions.clear();
+
             Page pageSplit;
             if (partitionSize == page.getPositionCount()) {
-                pageSplit = page; // entire page will be sent to this partition, no copies necessary
+                // entire page will be sent to this partition, just compact the page to avoid over-retaining
+                // memory and match the behavior of the sub-partitioned case
+                page.compact();
+                pageSplit = page;
             }
             else {
                 pageSplit = page.copyPositions(positions.elements(), 0, partitionSize);
             }
-            // clear the assigned positions list for the next iteration to start empty
```

Comment:
I think it's a bug. You use `positions.elements()` after you run `positions.clear();`

Reply:
Internally `IntArrayList.clear()` just sets `IntArrayList#size = 0;` but does not modify the `int[] elements` array. Reseting the size to `0` immediately after the `if (size == 0)` check simplifies the logic once the single partition case is refactored to return early (in the next PR) since otherwise you would have to ensure the `clear()` call happens in both branches.

Reply:
> Internally IntArrayList.clear() just sets IntArrayList#size = 0; but does not modify the int[] elements array

I know, but it doesn't look good and you don't know if implementation of `clear` won't change in the future.